src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,575 @@
+            return Err(InvalidSubgraph::NotConvex);
+        }
+        let nodes = subpg.nodes_iter().map_into().collect_vec();
+        if nodes.is_empty() {
+            return Err(InvalidSubgraph::EmptySubgraph);
+        }
+        if !nodes.iter().map(|&n| base.get_parent(n)).all_equal() {
+            return Err(InvalidSubgraph::NoSharedParent);
+        }
+        Ok(Self { base, nodes })
+    }
+
+    /// Create a new convex sibling subgraph from a set of nodes.
+    ///
+    /// This fails if the set of nodes is not convex, nodes do not share a
+    /// common parent or the subgraph is empty.
+    pub fn try_new(base: &'g Base, nodes: Vec<Node>) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let mut checker = ConvexChecker::new(base.portgraph());
+        Self::try_new_with_checker(base, nodes, &mut checker)
+    }
+
+    /// Create a new convex sibling subgraph from a set of nodes.
+    ///
+    /// Provide a [`ConvexChecker`] instance to avoid constructing one for
+    /// faster convexity check. If you do not have one, use [`SiblingSubgraph::try_new`].
+    ///
+    /// This fails if the set of nodes is not convex, nodes do not share a
+    /// common parent or the subgraph is empty.
+    pub fn try_new_with_checker(
+        base: &'g Base,
+        nodes: Vec<Node>,
+        checker: &mut ConvexChecker<&'g Base::Portgraph>,
+    ) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        if !checker.is_node_convex(nodes.iter().map(|n| n.index)) {
+            return Err(InvalidSubgraph::NotConvex);
+        }
+        if nodes.is_empty() {
+            return Err(InvalidSubgraph::EmptySubgraph);
+        }
+        if !nodes.iter().map(|&n| base.get_parent(n)).all_equal() {
+            return Err(InvalidSubgraph::NoSharedParent);
+        }
+        Ok(Self { base, nodes })
+    }
+
+    /// An iterator over the nodes in the subgraph.
+    pub fn nodes(&self) -> &[Node] {
+        &self.nodes
+    }
+
+    /// Whether a port is at the subgraph boundary.
+    fn is_boundary_port(&self, n: Node, p: Port) -> bool
+    where
+        Base: HugrView,
+    {
+        self.nodes.contains(&n)
+            && self
+                .base
+                .linked_ports(n, p)
+                .any(|(n, _)| !self.nodes.contains(&n))
+    }
+
+    /// An iterator of the incoming boundary ports.
+    fn incoming_ports(&self) -> impl Iterator<Item = (Node, Port)> + '_
+    where
+        Base: HugrView,
+    {
+        self.boundary_ports(Direction::Incoming)
+    }
+
+    /// An iterator of the outgoing boundary ports.
+    fn outgoing_ports(&self) -> impl Iterator<Item = (Node, Port)> + '_
+    where
+        Base: HugrView,
+    {
+        self.boundary_ports(Direction::Outgoing)
+    }
+
+    /// An iterator of the boundary ports, either incoming or outgoing.
+    fn boundary_ports(&self, dir: Direction) -> impl Iterator<Item = (Node, Port)> + '_
```

Comment:
You might want to make some of these public. (If you keep them internal, you might want to just inline incoming/outgoing, but I think making those public is reasonable!)

Reply:
Man, you've got so many good points!

Thanks a lot for your thorough reviewing, an @acl-cqc reviewed code is just so much better than what I started with